core/src/cluster_slots_service/cluster_slots.rs

```diff
@@ -83,36 +119,106 @@ impl ClusterSlots {
         cluster_slots.get(idx as usize)
     }
 
-    pub(crate) fn update(
-        &self,
-        root_slot: Slot,
-        validator_stakes: &Arc<ValidatorStakesMap>,
-        cluster_info: &ClusterInfo,
-        root_epoch: u64,
-    ) {
-        let current_slot = self.current_slot.load(Ordering::Relaxed);
+    #[inline]
+    fn get_row_for_slot_mut(
+        slot: Slot,
+        cluster_slots: &mut VecDeque<RowContent>,
+    ) -> Option<&mut RowContent> {
+        let start = cluster_slots.front()?.slot;
+        if slot < start {
+            return None;
+        }
+        let idx = slot - start;
+        cluster_slots.get_mut(idx as usize)
+    }
+
+    pub(crate) fn update(&self, root_bank: &Bank, cluster_info: &ClusterInfo) {
+        let root_slot = root_bank.slot();
+        let current_slot = self.get_current_slot();
         if current_slot > root_slot + 1 {
             error!("Invalid update call to ClusterSlots, can not roll time backwards!");
             return;
         }
-        self.maybe_update_validator_stakes(validator_stakes, root_epoch);
+        let my_epoch = self.get_epoch_for_slot(self.get_current_slot());
+        let root_epoch = root_bank.epoch();
+        if self.need_to_update_epoch(root_epoch) {
+            self.update_epoch_info(my_epoch, root_bank);
+        }
+
         let epoch_slots = {
             let mut cursor = self.cursor.lock().unwrap();
             cluster_info.get_epoch_slots(&mut cursor)
         };
-        self.update_internal(root_slot, validator_stakes, epoch_slots);
-        self.report_cluster_slots_perf_stats();
+        self.update_internal(root_slot, epoch_slots);
+        self.maybe_report_cluster_slots_perf_stats();
+    }
+
+    fn need_to_update_epoch(&self, root_epoch: Epoch) -> bool {
+        let rg = self.root_epoch.read().unwrap();
+        let my_epoch = rg.as_ref().map(|v| v.number);
+        Some(root_epoch) != my_epoch
+    }
+
+    // call this to update internal datastructures for current and next epoch
+    fn update_epoch_info(&self, my_epoch: Option<Epoch>, root_bank: &Bank) {
+        let root_epoch = root_bank.epoch();
+        let epoch_stakes_map = root_bank.epoch_stakes_map();
+        let mut epoch_metadata = self.epoch_metadata.write().unwrap();
+        if let Some(my_epoch) = my_epoch {
+            epoch_metadata.remove(&my_epoch);
+        }
+        // Next we fetch info about current and upcoming epoch's stakes
+        epoch_metadata.insert(
+            root_epoch,
+            EpochStakeInfo::from(&epoch_stakes_map[&root_epoch]),
+        );
+        let next_epoch = root_epoch.wrapping_add(1);
+        let next_epoch_info = EpochStakeInfo::from(&epoch_stakes_map[&next_epoch]);
+        *self.root_epoch.write().unwrap() = Some(RootEpoch {
+            schedule: root_bank.epoch_schedule().clone(),
+            number: root_epoch,
+        });
+        let mut first_slot = root_bank
```

Comment:
We may need to patch our ringbuffer in case where epochs are short and our ringbuffer is actually longer than 2 epochs. This can happen near genesis.

Reply:
Yes, some of the tests in our CI have these conditions (I believe local_cluster).